test/extensions/wasm/test_data/stats_cpp.cc

```diff
@@ -0,0 +1,73 @@
+// NOLINT(namespace-envoy)
+#include <string>
+
+#include "proxy_wasm_intrinsics.h"
+
+// Test the low level interface.
+extern "C" EMSCRIPTEN_KEEPALIVE void proxy_onStart() {
+  auto c = defineMetric(MetricType::Counter, "test_counter");
+  auto g = defineMetric(MetricType::Gauge, "test_gauges");
+  auto h = defineMetric(MetricType::Histogram, "test_histogram");
+
+  incrementMetric(c, 1);
+  recordMetric(g, 2);
+  recordMetric(h, 3);
+
+  logTrace(std::string("get counter = ") + std::to_string(getMetric(c)));
+  incrementMetric(c, 1);
+  logDebug(std::string("get counter = ") + std::to_string(getMetric(c)));
+  recordMetric(c, 3);
+  logInfo(std::string("get counter = ") + std::to_string(getMetric(c)));
+  logWarn(std::string("get gauge = ") + std::to_string(getMetric(g)));
+  logError(std::string("get histogram = ") + std::to_string(getMetric(h)));
+}
+
+// Test the higher level interface.
+extern "C" EMSCRIPTEN_KEEPALIVE void proxy_onTick() {
+  auto c = new Metric(MetricType::Counter, "test_counter",
+                      {MetricTag{"counter_tag", MetricTag::TagType::String}});
+  auto g = new Metric(MetricType::Gauge, "test_gauge",
+                      {MetricTag{"gauge_int_tag", MetricTag::TagType::Int}});
+  auto h = new Metric(MetricType::Gauge, "test_histogram",
+                      {MetricTag{"histogram_int_tag", MetricTag::TagType::Int},
+                       MetricTag{"histogram_string_tag", MetricTag::TagType::String},
+                       MetricTag{"histogram_bool_tag", MetricTag::TagType::Bool}});
+
+  c->increment(1, "test_tag");
+  g->record(2, 9);
+  h->record(3, 7, "test_tag", true);
+
+  logTrace(std::string("get counter = ") + std::to_string(c->get("test_tag")));
+  c->increment(1, "test_tag");
+  logDebug(std::string("get counter = ") + std::to_string(c->get("test_tag")));
+  c->record(3, "test_tag");
+  logInfo(std::string("get counter = ") + std::to_string(c->get("test_tag")));
+  logWarn(std::string("get gauge = ") + std::to_string(g->get(9)));
+
+  auto h_id = h->resolve(7, "test_tag", true);
+  logError(std::string("resolved histogram name = ") + h->nameFromIdSlow(h_id));
+}
+
+// Test the high level interface.
+extern "C" EMSCRIPTEN_KEEPALIVE void proxy_onLog(uint32_t /* context_zero */) {
+  auto c =
+      Counter<std::string, int, bool>::New("test_counter", "string_tag", "int_tag", "bool_tag");
+  auto g = Counter<std::string, std::string>::New("test_counter", "string_tag1", "string_tag2");
+  auto h =
+      Counter<int, std::string, bool>::New("test_histogram", "int_tag", "string_tag", "bool_tag");
+
+  c->increment(1, "test_tag", 7, true);
+  logTrace(std::string("get counter = ") + std::to_string(c->get("test_tag", 7, true)));
+  auto simple_c = c->resolve("test_tag", 7, true);
+  simple_c++;
```

Comment:
Is this remaining in the worker thread and send into the envoy singleton periodically, or is every call going into the underlying envoy scope?  If it is the latter it may cause a lot of contention especially at ingress where there may be many worker thread.

Reply:
This is a test for the wasm "service".  This is a singleton.  It isn't used with any of the per-stream stuff, but just background activity, so it shouldn't contend.  There are no calls going into it except the Timer which is going to be called on the single thread that the singleton is bound to.